src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -0,0 +1,823 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import <os/lock.h>
+
+#import "MTRAsyncCallbackWorkQueue_Internal.h"
+#import "MTRBaseDevice_Internal.h"
+#import "MTRCluster.h"
+#import "MTRDeviceController_Internal.h"
+#import "MTRDevice_Internal.h"
+#import "MTRError_Internal.h"
+#import "MTREventTLVValueDecoder_Internal.h"
+#import "MTRLogging.h"
+
+#include "lib/core/CHIPError.h"
+#include "lib/core/DataModelTypes.h"
+#include <app/ConcreteAttributePath.h>
+
+#include <app/AttributePathParams.h>
+#include <app/BufferedReadCallback.h>
+#include <app/ClusterStateCache.h>
+#include <app/InteractionModelEngine.h>
+#include <platform/PlatformManager.h>
+
+typedef void (^MTRDeviceAttributeReportHandler)(NSArray * _Nonnull);
+
+// Consider moving utility classes to their own file
+#pragma mark - Utility Classes
+@interface MTRPair<FirstObjectType, SecondObjectType> : NSObject
++ (instancetype)pairWithFirst:(FirstObjectType)first second:(SecondObjectType)second;
+@property (nonatomic, readonly) FirstObjectType first;
+@property (nonatomic, readonly) SecondObjectType second;
+@end
+
+@implementation MTRPair
+- (instancetype)initWithFirst:(id)first second:(id)second
+{
+    if (self = [super init]) {
+        _first = first;
+        _second = second;
+    }
+    return self;
+}
++ (instancetype)pairWithFirst:(id)first second:(id)second
+{
+    return [[MTRPair alloc] initWithFirst:first second:second];
+}
+@end
+
+@interface MTRWeakReference<ObjectType> : NSObject
++ (instancetype)weakReferenceWithObject:(ObjectType)object;
+- (instancetype)initWithObject:(ObjectType)object;
+- (ObjectType)strongObject; // returns strong object or NULL
+@end
+
+@interface MTRWeakReference () {
+@private
+    __weak id _object;
+}
+@end
+
+@implementation MTRWeakReference
+- (instancetype)initWithObject:(id)object
+{
+    if (self = [super init]) {
+        _object = object;
+    }
+    return self;
+}
++ (instancetype)weakReferenceWithObject:(id)object
+{
+    return [[MTRWeakReference alloc] initWithObject:object];
+}
+- (id)strongObject
+{
+    return _object;
+}
+@end
+
+#pragma mark - SubscriptionCallback class declaration
+using namespace chip;
+using namespace chip::app;
+using namespace chip::Protocols::InteractionModel;
+
+typedef void (^DataReportCallback)(NSArray * value);
+typedef void (^ErrorCallback)(NSError * error);
+
+namespace {
+
+class SubscriptionCallback final : public ClusterStateCache::Callback {
```

Comment:
This is basically a copy/paste of the MTRBaseDevice code but converting the output to a different data structure?

It would be really good to share the shared bits if we can manage it (e.g. by having a superclass with most of the machinery and subclasses that do the data conversions).